internal/server/handlers/handlers.go

```diff
@@ -1,106 +1,216 @@
+		mh.l.Error().Err(err).Msg("error creating Metric")
+
+		return nil, errors.Wrap(err, "error creating new Metric")
+	}
+
+	return newMetric, nil
+}
+
+func (mh *MetricHandler) Retrieve(response http.ResponseWriter, request *http.Request) {
+	metricName := domain.MetricName(chi.URLParam(request, "metricName"))
+
+	metricModel, found := mh.st.Get(metricName)
+	if !found {
+		log.Error().Msgf("Retrieve: Failed to retrieve model from memory %s", metricName)
+		http.Error(response, "Failed to retrieve model from memory", http.StatusNotFound)
+
+		return
+	}
+
+	_, _ = response.Write([]byte(metricModel.GetValue()))
+	response.WriteHeader(http.StatusOK)
+}
+
+func (mh *MetricHandler) RetrieveModel(response http.ResponseWriter, request *http.Request) {
+	// Decode the incoming JSON request into the Metric struct
+	var metric *model.Metric
+	if err := json.NewDecoder(request.Body).Decode(&metric); err != nil {
+		mh.l.Error().Err(err).Msg("Invalid JSON input")
+		http.Error(response, "Invalid JSON input", http.StatusNotFound)
+
+		return
+	}
+
+	// Prepare the updated metric to be returned
+	responseMetric, found := mh.st.Get(metric.ID)
+	if !found {
+		mh.l.Error().Msgf("Failed to retrieve model from memory %s", metric.ID)
+		http.Error(response, "Failed to retrieve model from memory", http.StatusNotFound)
+
+		return
+	}
+
+	response.WriteHeader(http.StatusOK)
+	err := json.NewEncoder(response).Encode(responseMetric)
+	if err != nil {
+		mh.l.Error().Err(err).Msg("Failed to encode response JSON")
+		http.Error(response, "Failed to process response", http.StatusInternalServerError)
 	}
 }
 
-func (mh *MetricHandler) Render(w http.ResponseWriter, _ *http.Request) {
+func (mh *MetricHandler) Render(response http.ResponseWriter, _ *http.Request) {
 	page := struct {
-		Gauges   map[domain.MetricName]float64
-		Counters map[domain.MetricName]int64
+		Metrics map[domain.MetricName]model.Metric
 	}{
-		Gauges:   mh.st.GetGauges(),
-		Counters: mh.st.GetCounters(),
+		Metrics: mh.st.GetAll(),
 	}
 
-	reader := templates.NewEmbedReader()
+	reader := web.NewEmbedReader()
 	tmpl, err := reader.Read("index.html")
 	if err != nil {
-		http.Error(w, "Failed to load template: "+err.Error(), http.StatusInternalServerError)
+		mh.l.Error().Err(err).Msg("Could not load template")
+		http.Error(response, "Failed to load template: "+err.Error(), http.StatusInternalServerError)
+
 		return
 	}
 
-	if err := tmpl.Execute(w, page); err != nil {
-		http.Error(w, "Failed to render page: "+err.Error(), http.StatusInternalServerError)
+	if err := tmpl.Execute(response, page); err != nil {
+		mh.l.Error().Err(err).Msg("Could not render template")
+		http.Error(response, "Failed to render page: "+err.Error(), http.StatusInternalServerError)
 	}
 }
 
-// Embedding middleware setup in the constructor
-func (mh *MetricHandler) setRouter() {
-	mh.router.Use(middleware.Logger)
+// SetRouter Embedding middleware setup in the constructor.
+func (mh *MetricHandler) SetRouter() {
```

Comment:
[SUGGESTION] It should not be a part of `Handler`

Reply:
fixed